opentelemetry-api/src/opentelemetry/metrics/__init__.py

```diff
@@ -0,0 +1,343 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=too-many-ancestors
+# type: ignore
+
+# FIXME enhance the documentation of this module
+"""
+This module provides abstract and concrete (but noop) classes that can be used
+to generate metrics.
+"""
+
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+from os import environ
+from threading import RLock
+from typing import Optional, cast
+
+from opentelemetry.environment_variables import OTEL_PYTHON_METER_PROVIDER
+from opentelemetry.metrics.instrument import (
+    Counter,
+    DefaultCounter,
+    DefaultHistogram,
+    DefaultObservableCounter,
+    DefaultObservableGauge,
+    DefaultObservableUpDownCounter,
+    DefaultUpDownCounter,
+    Histogram,
+    ObservableCounter,
+    ObservableGauge,
+    ObservableUpDownCounter,
+    UpDownCounter,
+)
+from opentelemetry.util._providers import _load_provider
+
+_logger = getLogger(__name__)
+
+
+class MeterProvider(ABC):
+    def __init__(self):
+        super().__init__()
+        self._lock = RLock()
+
+    @abstractmethod
+    def get_meter(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ) -> "Meter":
+        if name is None or name == "":
+            _logger.warning("Invalid name: %s", name)
+
+
+class _DefaultMeterProvider(MeterProvider):
+    def get_meter(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ) -> "Meter":
+        with self._lock:
+            super().get_meter(name, version=version, schema_url=schema_url)
+            # This is done in order to make it possible to store configuration
+            # in the meter provider and make it automatically accessible for
+            # any meter even after it changes.
+            meter = _DefaultMeter(name, version=version, schema_url=schema_url)
+            meter._meter_provider = self  # pylint: disable=protected-access
```

Comment:
On second look, we don't use the `_meter_provider` field at all and I don't think it's needed. I guess the proxy provider covers this use case for us.

Reply:
The comment above explains why this attribute is being added, it specifically required [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#get-a-meter):

_This could, for example, be implemented by storing any mutable configuration in the MeterProvider and having Meter implementation objects have a reference to the MeterProvider from which they were obtained._

Reply:
That attribute is there to provide access to the `MeterProvider` to `Meter` instances as the spec says. It is not being used now, but it is still required. I don't really understand how does `ProxyMeterProvider` accomplishes that goal. Even if that was the case if `set_meter_provider` is called before `get_meter_provider` a `ProxyMeterProvider` won't be used.

Reply:
I want to point out one thing that we don't do anything like this in tracing(tracers don't have any reference of traceprovider they were obtained from). Since the metrics spec draws inspiration from trace spec we can probably not implement this concern and go with our earlier decision of  _by allowing to work with an outdated configuration_. That will also be consistent with our tracing.

Reply:
After asking about this in the metrics SIG I prefer to remove this for the time being.